lib/compilers/ada.js

```diff
@@ -53,91 +53,112 @@ export class AdaCompiler extends BaseCompiler {
-        // It's used in runCompiler to insert the input filename at the correct
-        // location in the command line.
-        // input filename is inserted before -cargs.
+        gnatmake_opts.push(
+            '-g',
+            '-fdiagnostics-color=always',
+            '-eS', // output commands to stdout, they are not errors
+            inputFilename,
+        );
 
         if (!filters.binary) {
-            // produce assembly output in outputFilename
-            options.push(
-                'compile',
-                '-g', // enable debugging
+            gnatmake_opts.push(
                 '-S', // Generate ASM
-                '-fdiagnostics-color=always',
-                '-fverbose-asm', // Generate verbose ASM showing variables
                 '-c', // Compile only
-                '-eS', // commands are not errors
-                '-cargs', // Compiler Switches for gcc.
+                '-fverbose-asm', // Generate verbose ASM showing variables
+            );
+
+            // produce assembly output in outputFilename
+            compiler_opts.push(
                 '-o',
                 outputFilename,
             );
 
+
             if (this.compiler.intelAsm && filters.intel) {
-                options = options.concat(this.compiler.intelAsm.split(' '));
+                gnatmake_opts.push(this.compiler.intelAsm.split(' '));
             }
         } else {
-            // produce an executable in the file specified in getExecutableFilename.
-            // The output file is automatically decided by GNAT based on the Source_File_Name being used.
-            // As we are for forcing 'example.adb', the output here will be 'example'.
-            options.push(
-                'make',
-                '-eS',
-                '-g',
-                '-cargs', // Compiler Switches for gcc.
+            gnatmake_opts.push(
+                '-o',
+                outputFilename,
             );
         }
 
-        return options;
+        // Spread the options coming from outside (user input or backend options)
+        // in the correct list.
+
+        let part = 0; // 0: gnatmake, 1: compiler, 2: linker, 3: binder
+        for (const a of backend_opts.concat(utils.splitArguments(this.compiler.options))) {
+            if (a === '-cargs') {
+                part = 1;
+                continue;
+            } else if (a === '-largs') {
+                part = 2;
+                continue;
+            } else if (a === '-bargs') {
+                part = 3;
+                continue;
+            } else if (a === '-margs') {
+                part = 0;
+                continue;
+            }
+
+            if (part == 0) {
+                gnatmake_opts.push(a);
+            } else if (part == 1) {
+                compiler_opts.push(a);
+            } else if (part == 2) {
+                linker_opts.push(a);
+            } else if (part ==_3) {
```

Comment:
`_3` typo?

Reply:
Yes... Don't know why the linter catches this here and not locally :/